### PR TITLE
VX_RPP Extend OpenVX tensor support for augmentations in RPP

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Extend OpenVX tensor support for 26 augmentations.
- Minor code clean up in Brightness, CMN, Resize, Copy, Nop and SequenceRearrange.

### CODE REVIEW

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h [hunk 1]
```diff
@@ -209,14 +236,41 @@ vx_status Resize_Register(vx_context);
 #define VX_KERNEL_RPP_CROPPD_NAME   							"org.rpp.CropPD"
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: @fiona-gladwin Please check indentation here and below

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 2]
```diff
@@ -982,9 +982,11 @@ extern "C"
 	/*! \brief [Graph] Creates a RPP Brightness function node.
 	 * \ingroup group_amd_rpp
 	 * \param [in] graph The handle to the graph.
-	 * \param [in] pSrc The input tensor data.
+	 * \param [in] pSrc The input tensor in <tt>\ref VX_TYPE_UINT8<tt> or <tt>\ref VX_TYPE_FLOAT32<tt> or 
+	 * <tt>\ref VX_TYPE_FLOAT16<tt> or <tt>\ref VX_TYPE_INT8<tt> format data.
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: @fiona-gladwin why the extra line *. This might cause issues with Doxygen. Keep all comments for on parameter as one continuous line

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 3]
```diff
@@ -982,9 +982,11 @@ extern "C"
...
 	 * \param [in] pSrcRoi The input tensor of batch size in <tt>unsigned int<tt> containing the roi values for the input in xywh/ltrb format.
-	 * \param [out] pDst The output tensor data.
+	 * \param [out] pDst The output tensor in <tt>\ref VX_TYPE_UINT8<tt> or <tt>\ref VX_TYPE_FLOAT32<tt> or 
+	 * <tt>\ref VX_TYPE_FLOAT16<tt> or <tt>\ref VX_TYPE_INT8<tt> format data.
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: @fiona-gladwin same as above. I see many more like this below too

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 4]
```diff
@@ -1005,27 +1007,29 @@ extern "C"
...
 	 * \param [in] roiType The type of roi <tt>\ref VX_TYPE_INT32<tt> denotes whether source roi is of XYWH/LTRB type.
 	 * \return <tt> vx_node</tt>.
 	 * \returns A node reference <tt>\ref vx_node</tt>. Any possible errors preventing a
 	 * successful creation should be checked using <tt>\ref vxGetStatus</tt>.
 	 */
-	SHARED_PUBLIC vx_node VX_API_CALL vxRppCropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_tensor pSrcRoi, vx_tensor pDst, vx_array pMultiplier, vx_array pOffset, vx_array pFlip, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType);
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppCropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_tensor pSrcRoi, vx_tensor pDst, vx_array pMultiplier, vx_array pOffset, vx_array pMirror, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType);
 
-	/*! \brief [Graph] Creates a RPP Nop function node.
+	/*! \brief [Graph] Creates a tensor Nop function node.
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: @fiona-gladwin Some of these have been changed from "RPP" to "tensor. Others haven't. Please be consistent

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_BUFFER_HIP, &data->pDst, sizeof(data->pDst)));
+#endif
+    } else if (data->deviceType == AGO_TARGET_AFFINITY_CPU) {
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[1], VX_TENSOR_BUFFER_HOST, &roi_tensor_ptr, sizeof(roi_tensor_ptr)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[0], VX_TENSOR_BUFFER_HOST, &data->pSrc, sizeof(data->pSrc)));
+        STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_BUFFER_HOST, &data->pDst, sizeof(data->pDst)));
+    }
+
+    data->pSrcRoi = reinterpret_cast<RpptROI *>(roi_tensor_ptr);
+    // Fill width and height array with ROI data required by RPP batchPD kernels
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:62]: @fiona-gladwin WHy is this going to call batch PD kernel?

File: amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    vx_status return_status = VX_SUCCESS;
+    BlurLocalData *data = NULL;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    refreshBlur(node, parameters, num, data);
+    if (data->deviceType == AGO_TARGET_AFFINITY_GPU) {
+#if ENABLE_OPENCL
+        return_status = VX_ERROR_NOT_IMPLEMENTED;
+#elif ENABLE_HIP
+        if (data->pSrcDesc->c == 1) {
+            rpp_status = rppi_blur_u8_pln1_batchPD_gpu(data->pSrc, data->pSrcDimensions, data->maxSrcDimensions, data->pDst, data->pKernelSize, data->pSrcDesc->n, data->handle->rppHandle);
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:129]: @fiona-gladwin why is it calling rppi function? isn't rppi = image? rppt = tensor?

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp (hunk 6 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:129]: Yes, but for certain augmentations like Blur, snow, rain there is no rppt calls present in RPP. So internally we will be calling the batchPD kernels for these augmentations

For augmentations like Pixelate, Vignette, Jitter and ColorTemperature the support is added for rppt but the codes are not pushed to TOT so it is commented for now

File: amd_openvx_extensions/amd_rpp/source/tensor/FishEye.cpp [hunk 7]
```diff
@@ -0,0 +1,252 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/source/tensor/FishEye.cpp]: @fiona-gladwin All copyright needs to be updated to say 2023, not 2019-2023

Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/FishEye.cpp]: Done

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h (hunk 1 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: @LakshmiKumar23 I am not able to see any indentation issues when I check in VScode please let me know which line the issue is

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 4 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: The Nop node does not call any RPP functions internally, same applies for copy and sequence rearrange hence we have named it as tensor Nop function node

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 8]
```diff
@@ -994,9 +996,9 @@ extern "C"
 	 * \returns A node reference <tt>\ref vx_node</tt>. Any possible errors preventing a
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Combine line #995 and #996 into one. They are both saying the same thing and will be better for doxygen.  This is true for all the fuctions.

Comment (LakshmiKumar23) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Yes, as mentioned in [this](https://github.com/LakshmiKumar23/MIVisionX/pull/32/files/897577f1145788898b0343d204be6e99728df33a#r1280922686) too

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 4 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Yes. Still we want the name and arguments to be consistant all throughout the API. The user of vx_rpp doesn't really care whether it calls a low level RPP or not

Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: please remove tensor prefix here

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 9]
```diff
@@ -1034,25 +1038,528 @@ extern "C"
...
 	 * successful creation should be checked using <tt>\ref vxGetStatus</tt>.
 	 */
-	SHARED_PUBLIC vx_node VX_API_CALL vxRppNop(vx_graph graph, vx_tensor pSrc, vx_tensor pDst);
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppNop(vx_graph graph, vx_tensor pSrc, vx_tensor pDst);
 
-	/*! \brief [Graph] Creates a RPP vxRppResize function node.
+	/*! \brief [Graph] Creates a RPP Resize function node.
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: please remove RPP prefix here

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h [hunk 10]
```diff
@@ -1034,25 +1038,528 @@ extern "C"
...
+	 * \param [out] pDst The output tensor <tt>\ref VX_TYPE_UINT8<tt> format data.
+	 * \param [in] pNewOrder The rearrange order in <tt>\ref VX_TYPE_UINT32<tt> containing the order in which frames are copied.
+	 * \param [in] layout The layout in <tt>\ref VX_TYPE_INT32<tt> denotes the layout of input and output tensor.
+	 * \return <tt> vx_node</tt>.
+	 * \returns A node reference <tt>\ref vx_node</tt>. Any possible errors preventing a
+	 * successful creation should be checked using <tt>\ref vxGetStatus</tt>.
+	 */
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppSequenceRearrange(vx_graph graph, vx_tensor pSrc, vx_tensor pDst, vx_array pNewOrder, vx_scalar layout);
+
+	/*! \brief [Graph] Creates a RPP Blend function node.
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: I think we need to remove RPP prefix from all the descriptions since it is implied. We don't want to describe whether it is calling low level function or not

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h (hunk 1 above)
Comment (LakshmiKumar23) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: @fiona-gladwin Please open the file on git and check. It might be space vs tab and looks same on vs code

File: amd_openvx_extensions/amd_rpp/source/image/SequenceRearrangebatchPD.cpp [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    RppiSize dimensions;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32u device_type;
+    vx_uint32 new_sequence_length;
+    vx_uint32 sequence_length;
+    vx_uint32 sequence_count;
+    vx_uint32 *new_order;
+#if ENABLE_OPENCL
+    cl_mem cl_pSrc;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/image/SequenceRearrangebatchPD.cpp]: I think I have commented on this before. We shouldn't mix camel and snake_case both. Please keep everything in single type

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: amd_openvx_extensions/amd_rpp/source/image/SequenceRearrangebatchPD.cpp (hunk 11 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/image/SequenceRearrangebatchPD.cpp]: This is a batchPD file. In our previous PR we made tensor changes in the batchPD code itself. So we have have reintroduced the file and changed name alone. In all our tensor files we have followed the naming convention in all files

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 10 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Removed for all tensor API

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 9 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 4 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 8 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 4 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 3 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done

File: amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h (hunk 2 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h]: Done. changed into single line

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h (hunk 1 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: In git file also I am unable to see the issue

**Comment by fiona-gladwin** [comment]
@rrawther and @LakshmiKumar23 Addressed all review comments
As discussed. Modified the Doxygen description of all tensor API in this PR.

**Comment by LakshmiKumar23** [comment]
> @rrawther and @LakshmiKumar23 Addressed all review comments As discussed. Modified the Doxygen description of all tensor API in this PR.

@fiona-gladwin did the document build successfully?

**Comment by fiona-gladwin** [comment]
> > @rrawther and @LakshmiKumar23 Addressed all review comments As discussed. Modified the Doxygen description of all tensor API in this PR.
> 
> @fiona-gladwin did the document build successfully?

Yes @LakshmiKumar23 I was able to build the documentation, but with multiple warnings
```
====================== slowest reading durations =======================
0.610 docs/docs/docs/README
0.539 README
0.534 docs/README
0.532 docs/docs/README
0.287 docs/docs/CHANGELOG
build succeeded, 2821 warnings.

The HTML pages are in _build/html.
```

File: amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    // Querying for output tensor
+    data->pDstDesc = new RpptDesc;
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[3], VX_TENSOR_NUMBER_OF_DIMS, &data->pDstDesc->numDims, sizeof(data->pDstDesc->numDims)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[3], VX_TENSOR_DIMS, &data->ouputTensorDims, sizeof(vx_size) * data->pDstDesc->numDims));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[3], VX_TENSOR_DATA_TYPE, &output_tensor_dtype, sizeof(output_tensor_dtype)));
+    data->pDstDesc->dataType = getRpptDataType(output_tensor_dtype);
+    data->pDstDesc->offsetInBytes = 0;
+    fillDescriptionPtrfromDims(data->pDstDesc, data->outputLayout, data->ouputTensorDims);
+
+    data->pAlpha = static_cast<Rpp32f *>(malloc(sizeof(Rpp32f) * data->pSrcDesc->n));
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp]: please use new  and delete instead

File: amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    refreshBlend(node, parameters, num, data);
+    STATUS_ERROR_CHECK(createRPPHandle(node, &data->handle, data->pSrcDesc->n, data->deviceType));
+    STATUS_ERROR_CHECK(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeBlend(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    BlendLocalData *data;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->pAlpha != nullptr)  free(data->pAlpha);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp]: see comment above

File: amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_NUMBER_OF_DIMS, &data->pDstDesc->numDims, sizeof(data->pDstDesc->numDims)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_DIMS, &data->ouputTensorDims, sizeof(vx_size) * data->pDstDesc->numDims));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_DATA_TYPE, &output_tensor_dtype, sizeof(output_tensor_dtype)));
+    data->pDstDesc->dataType = getRpptDataType(output_tensor_dtype);
+    data->pDstDesc->offsetInBytes = 0;
+    fillDescriptionPtrfromDims(data->pDstDesc, data->outputLayout, data->ouputTensorDims);
+
+    data->maxSrcDimensions.height = data->pSrcDesc->h;
+    data->maxSrcDimensions.width = data->pSrcDesc->w;
+    data->pKernelSize = static_cast<Rpp32u *>(malloc(sizeof(Rpp32u) * data->pSrcDesc->n));
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp]: please use new instead of malloc whereever applicable

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp [hunk 15]
```diff
@@ -27,8 +27,8 @@ struct BrightnessLocalData {
     Rpp32u deviceType;
     RppPtr_t pSrc;
     RppPtr_t pDst;
-    vx_float32 *pAlpha;
-    vx_float32 *pBeta;
+    Rpp32f *pAlpha;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp]: why is this changed? It is better to use vx_ types for vx_extensions

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTemperature.cpp [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+    ColorTemperatureLocalData *data = new ColorTemperatureLocalData;
+    memset(data, 0, sizeof(ColorTemperatureLocalData));
+
+    vx_enum input_tensor_dtype, output_tensor_dtype;
+    int roi_type, input_layout, output_layout;
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[4], &input_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[5], &output_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[6], &roi_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[7], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    data->roiType = (roi_type == 0) ? RpptRoiType::XYWH : RpptRoiType::LTRB;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/ColorTemperature.cpp]: you can use better code here. Can't you pass the same value as enum for roiType and do a cast here?

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+    ColorTwistLocalData *data = new ColorTwistLocalData;
+    memset(data, 0, sizeof(ColorTwistLocalData));
+
+    vx_enum input_tensor_dtype, output_tensor_dtype;
+    int roi_type, input_layout, output_layout;
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[7], &input_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[8], &output_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[9], &roi_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[10], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    data->roiType = (roi_type == 0) ? RpptRoiType::XYWH : RpptRoiType::LTRB;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp]: please se the comment in  previous file. Applies to all

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+    RpptRoiType roiType;
+    vxTensorLayout inputLayout;
+    vxTensorLayout outputLayout;
+    size_t inputTensorDims[RPP_MAX_TENSOR_DIMS];
+    size_t outputTensorDims[RPP_MAX_TENSOR_DIMS];
+};
+
+static vx_status VX_CALLBACK refreshColorTwist(vx_node node, const vx_reference *parameters, vx_uint32 num, ColorTwistLocalData *data) {
+    vx_status status = VX_SUCCESS;
+    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[3], 0, data->inputTensorDims[0], sizeof(Rpp32f), data->pAlpha, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp]: I have mentioned this before. We need to pass this as a single tensor ptr with 4 float values for (alpha, beta, hue and sat). This was we can pass the same ptr (host/device) to RPP without copying

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+    // Querying for output tensor
+    data->pDstDesc = new RpptDesc;
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_NUMBER_OF_DIMS, &data->pDstDesc->numDims, sizeof(data->pDstDesc->numDims)));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2], VX_TENSOR_DIMS, &data->outputTensorDims, sizeof(vx_size) * data->pDstDesc->numDims));
+    STATUS_ERROR_CHECK(vxQueryTensor((vx_tensor)parameters[2],VX_TENSOR_DATA_TYPE, &output_tensor_dtype, sizeof(output_tensor_dtype)));
+    data->pDstDesc->dataType = getRpptDataType(output_tensor_dtype);
+    data->pDstDesc->offsetInBytes = 0;
+    fillDescriptionPtrfromDims(data->pDstDesc, data->outputLayout, data->outputTensorDims);
+
+    data->pAlpha = static_cast<Rpp32f *>(malloc(sizeof(Rpp32f) * data->pSrcDesc->n));
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp]: we need to pass the values directly from rocAL to rpp without copying as a single tensor

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp (hunk 15 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp]: Most of the members in this structure makes use of Rpp struct definitions and types. Hence in order to maintain uniformity and for better code we have changed all members to be of Rpp type. Please let us know if we can retain the same

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp (hunk 18 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp]: Hi @rrawther 
We discussed about the changes required for this in call and we decided not to do this.
Below were the pointers discussed.
Currently we have all of these variables as vx_array in rocAL as they are generated using the Parameter class. Internally it accepts batch size and generates random or fixed values and fills the vx_array.
In order to convert to vx_tensor we need to create a tensor fill with all the array values and again at the end when we need to pass to RPP we must split it into the 4 arrays.
Also in cases where we have multiple inputs with different data types we might not be able to do it.

File: amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp (hunk 12 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp]: Done, changed in all files

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTemperature.cpp (hunk 16 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/ColorTemperature.cpp]: Done, changed in all files

**Comment by fiona-gladwin** [comment]
@rrawther Addressed all review comments

File: amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp [hunk 20]
```diff
@@ -27,9 +27,9 @@ struct CropMirrorNormalizeLocalData {
     Rpp32u deviceType;
     RppPtr_t pSrc;
     RppPtr_t pDst;
-    vx_float32 *pMultiplier;
-    vx_float32 *pOffset;
-    vx_uint32 *pMirror;
+    Rpp32f *pMultiplier;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp]: I think it is better to use vx_types for OpenVX. Also we need to pass parameters through a single tensor instead of multiple vx_arrays

File: amd_openvx_extensions/amd_rpp/source/tensor/Fog.cpp [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+    refreshFog(node, parameters, num, data);
+    STATUS_ERROR_CHECK(createRPPHandle(node, &data->handle, data->pSrcDesc->n, data->deviceType));
+    STATUS_ERROR_CHECK(vxSetNodeAttribute(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    return VX_SUCCESS;
+}
+
+static vx_status VX_CALLBACK uninitializeFog(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    FogLocalData *data;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->pFogValue != nullptr)  free(data->pFogValue);
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Fog.cpp]: please use new and delete whereever possible

File: amd_openvx_extensions/amd_rpp/source/tensor/Glitch.cpp [hunk 22]
```diff
@@ -0,0 +1,270 @@
...
+*/
+
+#include "internal_publishKernels.h"
+
+struct GlitchLocalData {
+    vxRppHandle *handle;
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32u *pXOffsetR;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Glitch.cpp]: We need to combine all these offsets into a single tensor parameter

File: amd_openvx_extensions/amd_rpp/source/tensor/LensCorrection.cpp [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+*/
+
+#include "internal_publishKernels.h"
+
+struct LensCorrectionLocalData {
+    vxRppHandle *handle;
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32f *pStrength;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/LensCorrection.cpp]: I see this in all the kernels where parameters are passed as arrays which needs extra copies and reformatting in vx_rpp

File: amd_openvx_extensions/amd_rpp/source/tensor/Rain.cpp [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+*/
+
+#include "internal_publishKernels.h"
+
+struct RainLocalData {
+    vxRppHandle *handle;
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32f *pRainValue;
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Rain.cpp]: same as comment before

**Review by rrawther — CHANGES_REQUESTED** [review]
I think we need to change how the parameters are passed from rocAL to vx_rpp. Combining all the parameters into one tensor and using pinned memory will make it more efficient. I have discussed this before. But all the code is still using the batch_pd style

File: amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp (hunk 20 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp]: Changed to vx_types

**Comment by fiona-gladwin** [comment]
Changed all RPP types to vx data types.
As discussed will work on the vx_tensor changes in the future

File: amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp (hunk 15 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp]: Done

File: amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp (hunk 19 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp]: Please make a not of this to do it as a separate PR

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h (hunk 1 above)
Comment (LakshmiKumar23) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: @fiona-gladwin. Refer below. It's all over the place
image

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]

File: rocAL/rocAL/include/api/rocal_api_types.h [hunk 25]
```diff
@@ -177,8 +177,8 @@ enum RocalTensorBackend {
  * currently supports following formats
  */
 enum class RocalROICordsType {
-    ROCAL_XYWH = 0,
-    ROCAL_LTRB = 1
+    ROCAL_LTRB = 0,
+    ROCAL_XYWH = 1
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/api/rocal_api_types.h:181]: @fiona-gladwin Why is this interchanged? Has this been reflected in all use cases?

File: rocAL/rocAL/include/pipeline/commons.h [hunk 26]
```diff
@@ -106,8 +106,8 @@ enum class DecodeMode
  */
 enum class RocalROIType
 {
-    XYWH = 0,
-    LTRB
+    LTRB = 0,
+    XYWH
 };
```
Comment (LakshmiKumar23) [inline rocAL/rocAL/include/pipeline/commons.h:111]: @fiona-gladwin same as abaove?

File: rocAL/rocAL/include/api/rocal_api_types.h (hunk 25 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/api/rocal_api_types.h:181]: https://github.com/LakshmiKumar23/MIVisionX/pull/32#discussion_r1286480719
With respect to the above comment, We need to do a typecast from the RocalROIType to the RpptROIType directly
Hence the order should be the same as that of RPP. Hence this change is made

File: amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h (hunk 1 above)
Comment (fiona-gladwin) [reply amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h]: Done. Please check

**Comment by LakshmiKumar23** [comment]
@rrawther @fiona-gladwin ready to merge?
